### ISSUE TITLE
New plugin (token) for aria-live

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab#16145]
### Problem

Better screen-reader integration was requested/proposed in a few issues/pull requests:
- https://github.com/jupyterlab/jupyterlab/issues/4878
- https://github.com/jupyterlab/jupyterlab/issues/6573
- https://github.com/jupyterlab/jupyterlab/pull/15048
- https://github.com/jupyterlab/jupyterlab/pull/15554

However, there is repetitive code in #15048 and #15554 which writes directly to DOM elements to cause the side-effect of reading out text by screen reader. As highlighted in the reviews:
- this needs customization such as the delay for [#15554](https://github.com/jupyterlab/jupyterlab/pull/15554)
- the entries need to be cleaned at some timeout to avoid memory leak
- user needs to be able to toggle this feature off

### Proposed Solution

Create a new plugin with a public API to `announce()` a message to screen users, and with settings to provide customisability. 

### Additional context

None

**Comment by krassowski** [comment]
I'm not quite sure where this plugin should live. So far the aria-live region element was proposed to be added to the `shell` in `application` package, which is good in the sense that everyone gets access to it, but is not great for reusability (we would want to have it in Notebook to, but notebook has a different `application`).

A new package is an option, though it seems a bit slim, so maybe `apputils`?

**Comment by krassowski** [comment]
Note: cross-check with interface for jupyterlab logger and with the announcements API

**Comment by tonyfast** [comment]
i'm going to add some color from my experience with aria live in notebooks for all, nbconvert-a11y, and jupyter research. as you mentioned, this feature is similar to the jupyterlab logger and notifications. in fact, they may all be same thing. 

[this notebook describes an accessible, visual, and non-visual approach to logging](https://nbviewer.org/github/tonyfast/tonyfast/blob/a7a7697ecc741557324c5604b6b3b6f5f34e509d/tonyfast/xxiv/2024-07-11-activity-log.ipynb). in this approach, we use a persistent table that can be accessed visually, and provides the proper markup to interact with many of the nuances across assistive technology. as a landmark and table there are multiple ways to access the actions with assistive technology. a persistent table will contribute positively to the [wcag cognitive accessibility objective: ensure processes do not rely on memory](https://www.w3.org/WAI/WCAG2/supplemental/objectives/o6-memory/).

basically, the table of events allows toast notifications to persistent a little longer. extra long notifications could use the native notifications api. the activity log could be a persistent place for visual and non-visual to users to revisit their recent activity.

**Comment by IsabelParedes** [comment]
Hi! I would like to work on this. If you have any more information you can provide to approach this, I would highly appreciate it. I do have a couple of questions.

> A new package is an option, though it seems a bit slim, so maybe apputils?

Would it be OK to keep it in `apputils`?


> Note: cross-check with interface for jupyterlab logger and with the announcements API

And could you elaborate more on this?

Thank you!

**Comment by krassowski** [comment]
> Would it be OK to keep it in apputils?

Sure!

> > Note: cross-check with interface for jupyterlab logger and with the announcements API
> 
> And could you elaborate more on this?

I don't remember what I meant, but it would have been one or both:
- let's see what the method/argument names for the APIs are https://jupyterlab.readthedocs.io/en/latest/api/classes/apputils.NotificationManager.html and https://jupyterlab.readthedocs.io/en/latest/api/interfaces/logconsole.ILogger.html since these may have some good ideas (e.g. having a configurable level) and to avoid having completely separate naming conventions
- let's think whether e.g. error logs sent to logconsole and/or notification sent to notification manager should trigger an aria-live announcement; if that's the case, these plugins would need to depend on aria-live token
